apps/aecore/lib/aecore/oracle/oracle_state_tree.ex

```diff
@@ -0,0 +1,212 @@
+defmodule Aecore.Oracle.OracleStateTree do
+  @moduledoc """
+  Top level oracle state tree.
+  """
+  alias Aeutil.PatriciaMerkleTree
+  alias Aeutil.Serialization
+  alias Aecore.Oracle.Tx.OracleQueryTx
+  alias Aecore.Account.AccountStateTree
+
+  @type oracles_state :: %{oracle_tree: Trie.t(), oracle_cache_tree: Trie.t()}
+  @dummy_val <<0>>
+
+  @spec init_empty() :: oracles_state()
+  def init_empty do
+    %{
+      oracle_tree: PatriciaMerkleTree.new(:oracles),
+      oracle_cache_tree: PatriciaMerkleTree.new(:oracles_cache)
+    }
+  end
+
+  @spec prune(Chainstate.t(), non_neg_integer()) :: Chainstate.t()
+  def prune(chainstate, block_height) do
+    {new_oracles_state, new_accounts_state} =
+      initialize_deletion({chainstate.oracles, chainstate.accounts}, block_height - 1)
+
+    %{chainstate | oracles: new_oracles_state, accounts: new_accounts_state}
+  end
+
+  @spec enter_oracle(oracles_state(), map()) :: oracles_state()
+  def enter_oracle(tree, oracle) do
+    add_oracle(tree, oracle, :enter)
+  end
+
+  @spec insert_oracle(oracles_state(), map()) :: oracles_state()
+  def insert_oracle(tree, oracle) do
+    add_oracle(tree, oracle, :insert)
+  end
+
+  @spec get_oracle(oracles_state(), binary()) :: map()
+  def get_oracle(tree, key) do
+    get(tree.oracle_tree, key)
+  end
+
+  @spec exists_oracle?(oracles_state(), binary()) :: boolean()
+  def exists_oracle?(tree, key) do
+    exists?(tree, key, :oracle)
+  end
+
+  @spec enter_query(oracles_state(), map()) :: oracles_state()
+  def enter_query(tree, query) do
+    add_query(tree, query, :enter)
+  end
+
+  @spec insert_query(oracles_state(), map()) :: oracles_state()
+  def insert_query(tree, query) do
+    add_query(tree, query, :insert)
+  end
+
+  @spec get_query(oracles_state(), binary()) :: map()
+  def get_query(tree, key) do
+    get(tree.oracle_tree, key)
+  end
+
+  @spec exists_query?(oracles_state(), binary()) :: boolean()
+  def exists_query?(tree, key) do
+    exists?(tree, key, :oracle_query)
+  end
+
+  defp initialize_deletion({oracles_state, _accounts_state} = trees, expires) do
```

Comment:
This could get a better name, why do we need a separate function from prune anyways?

Reply:
@thepiwo 
Prune is API function and is implicitly - called function.
Initialize_deletion prepares the keys from the given tree and calls function filter_expired. In filter_expired function we filter and delete expired oracles and queries therefore we have 2 separate logics. That's why our code is segmented.

Reply:
okay, but as commented as well it would be good to split this logic from actual tree handling